arviz/data/inference_data.py

```diff
@@ -182,36 +211,103 @@ def __delattr__(self, group):
-            yield (group, getattr(self, group))
+        def __init__(self, parent: "InferenceData") -> None:
+            """Create a new InferenceDataValuesView from an InferenceData object."""
+            self.parent = parent
+
+        def __len__(self) -> int:
+            """Return the number of groups in the parent InferenceData."""
+            return len(self.parent._groups_all)
+
+        def __iter__(self) -> Iterator[xr.Dataset]:
+            """Iterate through the Xarray datasets present in the InferenceData object."""
+            parent = self.parent
+            for group in parent._groups_all:
+                yield getattr(parent, group)
+
+        def __contains__(self, key: object) -> bool:
+            """Return True if the given Xarray dataset is one of the values, and False otherwise."""
+            if not isinstance(key, xr.Dataset):
+                return False
+
+            for dataset in self:
+                if dataset.equals(key):
+                    return True
+
+            return False
+
+    def values(self) -> "InferenceData.InferenceDataValuesView":
+        """Return a view over the Xarray Datasets present in the InferenceData object."""
+        return InferenceData.InferenceDataValuesView(self)
+
+    class InferenceDataItemsView(ItemsView[str, xr.Dataset]):
+        """ItemsView implementation for InferenceData, to allow it to implement Mapping."""
+
+        def __init__(self, parent: "InferenceData") -> None:
+            """Create a new InferenceDataItemsView from an InferenceData object."""
+            self.parent = parent
+
+        def __len__(self) -> int:
+            """Return the number of groups in the parent InferenceData."""
+            return len(self.parent._groups_all)
+
+        def __iter__(self) -> Iterator[Tuple[str, xr.Dataset]]:
+            """Iterate through the groups and corresponding Xarray datasets in the InferenceData."""
+            parent = self.parent
+            for group in parent._groups_all:
+                yield group, getattr(parent, group)
+
+        def __contains__(self, key: object) -> bool:
+            """Return True if the (group, dataset) tuple is present, and False otherwise."""
+            parent = self.parent
+            if not isinstance(key, tuple) or len(key) != 2:
+                return False
+
+            group, dataset = key
+            if group not in parent._groups_all:
+                return False
+
+            if not isinstance(dataset, xr.Dataset):
+                return False
+
+            existing_dataset = getattr(parent, group)
+            return existing_dataset.equals(dataset)
+
+    def items(self) -> "InferenceData.InferenceDataItemsView":
+        """Return a view over the groups and datasets present in the InferenceData object."""
+        return InferenceData.InferenceDataItemsView(self)
 
     @staticmethod
-    def from_netcdf(filename):
+    def from_netcdf(filename: str) -> "InferenceData":
```

Comment:
Why sometimes strings are used for the types and sometimes the raw type? Trying to understand how typing works

Reply:
That's a good question. Type hints with string literals like `"InferenceData"` here represent so-called forward references: https://www.python.org/dev/peps/pep-0484/#forward-references

Forward references allow us to refer to types and values that are not currently in scope. On this particular line, the `from_netcdf` function is being defined within the `InferenceData` class, which means that the `InferenceData` class is not fully constructed yet i.e. the `InferenceData` name does not yet refer to anything in the current scope. If we attempted to use `InferenceData` here, it would cause a `NameError` because the symbol is not defined yet. Instead, if we use a forward reference, we signal to mypy that it should look up what that name refers to _when type-checking_ rather than when defining the class, which avoids the problem since by the time type-checking starts, the class has already been fully defined and its name can be resolved just fine.

This is why on [line 1334](https://github.com/arviz-devs/arviz/pull/1491/files#diff-c4c20d9d7fa9d08b420556e1e63dc4df0b417a78ceb0c52b036b2dd07791fc3fR1334) we can use `InferenceData` without the quotes: that line is located outside and after the `InferenceData` class definition, so that symbol is defined and present in the scope at the point where it is being used.

For `"Literal[True]"` and similar cases, we use quotes for a related but slightly different reason: the import of `Literal` on [line 42](https://github.com/arviz-devs/arviz/pull/1491/files#diff-c4c20d9d7fa9d08b420556e1e63dc4df0b417a78ceb0c52b036b2dd07791fc3fR42) is only performed if the `typing.TYPE_CHECKING` special value is true. However, type hints that are not forward references are always executed -- even when not performing type-checking. This means that if we had used `Literal[True]` instead of using the string literal equivalent, executing the `Literal[True]` expression in a non-type-checking run would cause a `NameError` since `Literal` is not imported in that situation.

Reply:
Thanks!